README.rst
===============================
Yet Another Murmurhash3 Binding
===============================

.. image:: https://img.shields.io/pypi/v/yammh3.svg
        :target: https://pypi.python.org/pypi/yammh3

.. image:: https://img.shields.io/travis/rolando/yammh3.svg
        :target: https://travis-ci.org/rolando/yammh3

.. image:: https://readthedocs.org/projects/yammh3/badge/?version=latest
        :target: https://readthedocs.org/projects/yammh3/?badge=latest
        :alt: Documentation Status


Python/Cython Murmurhash3 binding.

* Free software: MIT license
* Documentation: https://yammh3.readthedocs.org.

Features
--------

* Provides a high-level Python API.
* Provides a low-level Cython binding.
* Python 2 and 3 support.

Example
-------

Here is an example in Python:

.. code:: python

  from yammh3 import hash64

  key = b"yammh3!"

  # hash* functions return a signed integer by default.
  print("signed 64 bits hash is %s" % hash64(key))  # -> -1339990020854215562
  print("unsigned 64 bits hash is %s" % hash64(key, signed=False))  # -> 17106754052855336054L


In Cython, first we need to write a ``.pyx`` file with our code:

.. code:: cython

    # file: yammh3_example.pyx
    # mhash* functions are only available via cimport.
    from yammh3._yammh3 cimport mhash64, mhash64s
    from yammh3._yammh3 cimport int64_t, uint64_t, uint32_t

    def print_hashes(bytes key):
        cdef uint64_t h1
        cdef int64_t h2
        cdef uint32_t n = len(key)
        cdef char *c_key = <char *>key

        with nogil:  # releasing the GIL!
            h1 = mhash64(c_key, n)
            h2 = mhash64s(c_key, n)

        print("unsigned 64 bits hash is %d" % h1)
        print("signed 64 bits hash is %d" % h2)


We need to compile it as a module, usually by using a setup script:

.. code:: python

    # file: setup.py
    from setuptools import setup
    from setuptools.extension import Extension
    from Cython.Build import cythonize

    import yammh3  # already installed

    setup(
        name='yammh3-example',
        ext_modules=cythonize([
            Extension('*', ['*.pyx'], include_dirs=[yammh3.get_include()]),
        ])
    )


Then we build the modules in-place:

.. code::

    $ python setup.py build_ext --inplace
    Running build_ext
    building 'yammh3_example' extension
    ... [snip] ...
    copying build/lib.macosx-10.5-x86_64-2.7/yammh3_example.so ->


Now we are ready to run our code:

.. code::

    $ python -c 'import yammh3_example; yammh3_example.print_hashes(b"yammh3!")'
    unsigned 64 bits hash is 17106754052855336054
    signed 64 bits hash is -1339990020854215562


Credits
---------

Murmurhash3 was originally created by `Austin Appleby`_.

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _`Austin Appleby`: https://github.com/aappleby/smhasher


HISTORY.rst
=======
History
=======

0.2.0dev
------------------

* TODO


0.1.2 (2016-06-24)
------------------

* Documentation updates.

0.1.1 (2016-06-23)
------------------

* First **usable** release :)

0.1.0 (2016-06-17)
------------------

* First release on PyPI.


setup_tests.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import sys

from setuptools import setup
from setuptools.extension import Extension

from Cython.Build import cythonize

import yammh3


def get_extensions():
    kwargs = {
        'include_dirs': [yammh3.get_include()],
    }
    return cythonize([
        Extension('tests._test_cimports', ['tests/_test_cimports.pyx'], **kwargs),
    ])


setup(
    name='yammh3-tests',
    version='0.2.0dev',
    packages=['tests'],
    package_data={
        "tests": ['*.pyx', '*.pxd'],
    },
    ext_modules=get_extensions(),
)


tests/test_yammh3.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_yammh3
----------------------------------

Tests for `yammh3` module.
"""
import functools
import glob
import os
import pytest

from hypothesis import given, strategies as st
from yammh3 import get_include, hash128, hash64, hash32


def partial(func, **default_kwargs):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        kwargs = dict(default_kwargs, **kwargs)
        return func(*args, **kwargs)
    wrapper.default_kwargs = default_kwargs
    return wrapper


st_custom_seed = functools.partial(st.integers, 1, 2**32-1)

booleans = [False, True]

hash_funcs_defaults = [
    hash32, hash64, hash128,
]

hash_funcs_signed = [
    partial(hash32, signed=True),
] + [
    partial(hash64, signed=True, low=low, x64=x64)
    for low in booleans for x64 in booleans
] + [
    partial(hash128, signed=True, x64=True),
    partial(hash128, signed=True, x64=False),
]

hash_funcs_unsigned = [
    partial(hash32, signed=False),
] + [
    partial(hash64, signed=False, low=low, x64=x64)
    for low in booleans for x64 in booleans
] + [
    partial(hash128, signed=False, x64=True),
    partial(hash128, signed=False, x64=False),
]


hash_funcs_all = hash_funcs_signed + hash_funcs_unsigned


def is_hash128_out(out):
    return isinstance(out, tuple) and len(out) == 2


def assert_uint32(n):
    assert 0 <= n < 2**32


def assert_int32(n):
    assert -2**31 <= n < 2**31


def assert_uint64(n):
    assert 0 <= n < 2**64


def assert_int64(n):
    assert -2**63 <= n < 2**63


def assert_uint128(n):
    assert is_hash128_out(n)
    a, b = n
    assert_uint64(a)
    assert_uint64(b)


def assert_int128(n):
    assert is_hash128_out(n)
    a, b = n
    assert_int64(a)
    assert_int64(b)


def test_get_include():
    path = get_include()
    headers = glob.glob(os.path.join(path, '*.pxd'))
    assert len(headers) > 0


@pytest.mark.parametrize('hash_func', hash_funcs_defaults)
@given(value=st.text(), seed=st_custom_seed())
def test_hash_default_signed(hash_func, value, seed):
    out = hash_func(value, seed=seed)
    out_signed = hash_func(value, signed=True, seed=seed)
    assert out == out_signed


@pytest.mark.parametrize('hash_func', hash_funcs_all)
@given(value=st.text(), seed=st_custom_seed())
def test_hash_bytes(hash_func, value, seed):
    out_default = hash_func(value.encode('utf8'))
    out = hash_func(value, seed=seed)
    assert out != out_default


@pytest.mark.parametrize('hash_func', hash_funcs_all)
@given(value=st.text(), seed=st_custom_seed())
def test_hash_custom_seed(hash_func, value, seed):
    out_default = hash_func(value)
    out = hash_func(value, seed=seed)
    assert out != out_default


@pytest.mark.parametrize('hash_func', hash_funcs_unsigned)
@given(value=st.text(), seed=st_custom_seed())
def test_hash_unsigned(hash_func, value, seed):
    out = hash_func(value, seed=seed)
    if is_hash128_out(out):
        assert out[0] >= 0 and out[1] >= 0
    else:
        assert out >= 0


@pytest.mark.parametrize('hash_func', hash_funcs_signed)
def test_hash_signed(hash_func):
    # As we test random values, we stop once we find one negative.
    @given(value=st.text(min_size=1), seed=st_custom_seed())
    def find_negative(value, seed):
        out = hash_func(value, seed=seed)
        if is_hash128_out(out):
            out = min(*out)
        if out < 0:
            raise FoundNegative

    try:
        find_negative()
    except FoundNegative:
        pass
    else:
        assert False, "Unable to find negative hash"


@pytest.mark.parametrize('hash_assert', [
    (hash32, [assert_uint32, assert_int32]),
    (hash64, [assert_uint64, assert_int64]),
    (hash128, [assert_uint128, assert_int128]),
])
@given(value=st.text(), seed=st_custom_seed())
def test_hash_int_ranges(hash_assert, value, seed):
    hash_func, (assert_u, assert_s) = hash_assert
    assert_s(hash_func(value, seed=seed))  # default signed
    assert_u(hash_func(value, seed=seed, signed=False))


@pytest.mark.parametrize('x64', booleans)
@given(value=st.text(), seed=st_custom_seed())
def test_hash64_is_hash128(x64, value, seed):
    high = hash64(value, seed=seed, x64=x64)  # default low=False
    low = hash64(value, seed=seed, low=True, x64=x64)
    out = hash128(value, seed=seed, x64=x64)
    assert (high, low) == out


class FoundNegative(Exception):
    pass


tests/test_cimports.py
from hypothesis import given, strategies as st

from ._test_cimports import test_cimports as _test_cimports


@given(st.text())
def test_cimports(value):
    b_value = value.encode('utf-8')
    _test_cimports(b_value, len(b_value))


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'rolando/yammh3'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


src/yammh3/__init__.py
# -*- coding: utf-8 -*-
import six

from ._yammh3 import (
    mhash128, mhash128s,
    mhash128_x64, mhash128s_x64,
    mhash64, mhash64s, mhash64l, mhash64sl,
    mhash64_x64, mhash64s_x64, mhash64l_x64, mhash64sl_x64,
    mhash32, mhash32s,
)


__author__ = 'Rolando Espinoza'
__email__ = 'rolando at rmax.io'
__version__ = '0.2.0dev'

__all__ = ['get_include', 'hash128', 'hash64', 'hash32']


def get_include():
    """Returns the includes directory.

    Returns
    -------
    out : str
        Directory that contains header files.

    """
    import os
    return os.path.abspath(os.path.dirname(__file__))


def hash128(value, seed=0, signed=True, x64=False):
    """Returns a 128 bits hash for the given value.

    Parameters
    ----------
    value : unicode or str
        Value to be hashed.
    seed : int, optional
        Seed for the hash algorithm. Default is 0.
    signed : bool, optional
        If True, returns a signed integer, otherwise returns an unsigned
        integer. Default is True.
    x64 : bool, optional
        Whether to use x64 optimized hash functions. Default is False.

    Returns
    -------
    out : tuple
        A tuple of 64 bits integers (high, low).

    """
    value = _tobytes(value)
    if x64:
        choices = (mhash128_x64, mhash128s_x64)
    else:
        choices = (mhash128, mhash128s)
    out = choices[signed](value, len(value), seed)
    return out['high'], out['low']


def hash64(value, seed=0, signed=True, low=False, x64=False):
    """Returns a 64 bits hash for the given value.

    This function uses the 128 bits hash function internally and the ``low``
    parameter controls which part of bits are returned.

    Parameters
    ----------
    value : unicode or str
        Value to be hashed.
    seed : int, optional
        Seed for the hash algorithm. Default is 0.
    signed : bool, optional
        If True, returns a signed integer, otherwise returns an unsigned
        integer. Default is True.
    low : bool, optional
        Whether to returns low bits rather than high bits. Default is False.
    x64 : bool, optional
        Whether to use x64 optimized hash functions. Default is False.

    Returns
    -------
    out : int
        A 64 bits integer.

    """
    value = _tobytes(value)
    if low:
        choices = (
            (mhash64l, mhash64sl),
            (mhash64l_x64, mhash64sl_x64),
        )[x64]
    else:
        choices = (
            (mhash64, mhash64s),
            (mhash64_x64, mhash64s_x64),
        )[x64]
    return choices[signed](value, len(value), seed)


def hash32(value, seed=0, signed=True):
    """Returns a 32 bits hash for the given value.

    Parameters
    ----------
    value : unicode or str
        Value to be hashed.
    seed : int, optional
        Seed for the hash algorithm. Default is 0.
    signed : bool, optional
        If True, returns a signed integer, otherwise returns an unsigned
        integer. Default is True.

    Returns
    -------
    out : int
        A 32 bits integer.

    """
    value = _tobytes(value)
    func = (mhash32, mhash32s)[signed]
    return func(value, len(value), seed)


def _tobytes(s, encoding='utf-8'):
    if isinstance(s, six.binary_type):
        return s
    if isinstance(s, six.text_type):
        return s.encode(encoding)
    raise ValueError("Unsupported type '%s'" % type(s))
